router/dataplane_spec.gobra

```diff
@@ -372,211 +372,203 @@ pure func numInfoFields(seg1Len int, seg2Len int, seg3Len int) int {
 }
 
 ghost 
-requires seg1Len > 0
-requires seg2Len >= 0
-requires seg3Len >= 0
-requires currHF >= 0
-ensures res <= currHF
+requires 0 < seg1Len
+requires 0 <= seg2Len
+requires 0 <= seg3Len
+requires 0 <= currHF
+ensures  res <= currHF
 decreases
 pure func computePrevSegLen(currHF int, seg1Len int, seg2Len int, seg3Len int) (res int){
 	return seg1Len > currHF ? 0 : ((seg1Len + seg2Len) > currHF ? seg1Len : seg1Len + seg2Len)
 }
 
-//expects the raw bytes of whole path
 ghost
-requires start + 12 <= len(raw)
-requires 0 <= start
-requires acc(&raw[start+2], _) && acc(&raw[start+3], _) && acc(&raw[start+4], _) && acc(&raw[start+5], _)
+requires idx + 12 <= len(raw)
+requires 0 <= idx
+requires acc(&raw[idx+2], _) && acc(&raw[idx+3], _) && acc(&raw[idx+4], _) && acc(&raw[idx+5], _)
+ensures  len(res.HVF.MsgTerm_Hash_.MsgTerm_MPair_2.MsgTerm_L_) > 0
 decreases
-pure func getHopField(raw []byte, start int) io.IO_HF {
-	return let inif2 := binary.BigEndian.Uint16(raw[start+2:start+4]) in
-		let egif2 := binary.BigEndian.Uint16(raw[start+4:start+6]) in
+pure func getHopField(raw []byte, idx int, beta set[io.IO_msgterm], asid io.IO_as, ainfo io.IO_ainfo) (res io.IO_HF) {
+	return let inif2 := binary.BigEndian.Uint16(raw[idx+2:idx+4]) in
+		let egif2 := binary.BigEndian.Uint16(raw[idx+4:idx+6]) in
+		let op_inif2 := inif2 == 0 ? none[io.IO_ifs] : some(io.IO_ifs(inif2)) in
+		let op_egif2 := egif2 == 0 ? none[io.IO_ifs] : some(io.IO_ifs(egif2)) in
+		let ts := io.IO_msgterm(io.MsgTerm_Num{ainfo}) in
+		let l := io.IO_msgterm(io.MsgTerm_L{seq[io.IO_msgterm]{ts, io.if2term(op_inif2), io.if2term(op_egif2), 
+			io.IO_msgterm(io.MsgTerm_FS{beta})}}) in
+		let hvf := io.mac(io.macKey(io.asidToKey(asid)), l) in
 		io.IO_HF(io.IO_HF_{
-			InIF2 : inif2 == 0 ? none[io.IO_ifs] : some(io.IO_ifs(inif2)),
-			EgIF2 : egif2 == 0 ? none[io.IO_ifs] : some(io.IO_ifs(egif2)),
-			HVF : computeMsgTerm(raw, start),
+			InIF2 : op_inif2,
+			EgIF2 : op_egif2,
+			HVF : hvf,
 		})
 }
 
-//expects the raw bytes of whole path
 ghost 
-requires 4 + 8 * numINF + offset * 12 + 12 <= len(raw)
-requires prevSegLen >= 0 && offset - prevSegLen >= -1
-requires 0 < numINF && numINF < 4
-requires acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _)
-decreases offset - prevSegLen + 1
-pure func getSegPast(raw []byte, prevSegLen int, offset int, numINF int) seq[io.IO_HF]{
-	return offset - prevSegLen == -1 ? seq[io.IO_HF]{} : 
-		(unfolding acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _) in 
-			seq[io.IO_HF]{getHopField(raw,4 + 8 * numINF + offset*12)}) ++ getSegPast(raw, prevSegLen, offset-1, numINF)
+requires  0 <= offset
+requires  0 <= currHFIdx && currHFIdx <= segLen
+requires  offset + 12 * segLen <= len(raw)
+requires  acc(sl.AbsSlice_Bytes(raw, 0, len(raw)), _)
+ensures   len(res) == segLen - currHFIdx
+ensures   forall k int :: 0 <= k && k < len(res) ==> 
+	len(res[k].HVF.MsgTerm_Hash_.MsgTerm_MPair_2.MsgTerm_L_) > 0
```

Comment:
I think that we should pick the trigger explicitly here (`res[k]` may be a good choice)